library/pkcs11_client.c

```diff
@@ -0,0 +1,527 @@
+        return( MBEDTLS_ECDSA_MAX_SIG_LEN( ctx->bit_length ) );
+    default:
+        return( 0 );
+    }
+}
+
+static int pkcs11_sign( void *ctx_arg,
+                        mbedtls_md_type_t md_alg,
+                        const unsigned char *hash, size_t hash_len,
+                        unsigned char *sig, size_t *sig_len,
+                        int (*f_rng)(void *, unsigned char *, size_t),
+                        void *p_rng )
+{
+    mbedtls_pk_pkcs11_context_t *ctx = ctx_arg;
+    CK_RV rv;
+    CK_MECHANISM mechanism = {0, NULL_PTR, 0};
+    CK_ULONG ck_sig_len;
+
+    /* This function takes size_t arguments but the underlying layer
+       takes unsigned long. Either type may be smaller than the other.
+       Legitimate values won't overflow either type but we still need
+       to check for overflow for robustness. */
+    if( hash_len > (CK_ULONG)( -1 ) )
+        return( MBEDTLS_ERR_PK_BAD_INPUT_DATA );
+    (void) f_rng;
+    (void) p_rng;
+
+    switch( ctx->key_type )
+    {
+#if defined(MBEDTLS_ECDSA_C)
+    case MBEDTLS_PK_ECKEY:
+        ck_sig_len = MBEDTLS_ECDSA_MAX_SIG_LEN( ctx->bit_length );
+        mechanism.mechanism = CKM_ECDSA;
+        break;
+#endif /* MBEDTLS_ECDSA_C */
+    default:
+        return( MBEDTLS_ERR_PK_UNKNOWN_PK_ALG );
+    }
+
+    rv = C_SignInit( ctx->hSession, &mechanism, ctx->hPrivateKey );
+    if( rv != CKR_OK )
+        goto exit;
+    rv = C_Sign( ctx->hSession, (CK_BYTE_PTR) hash, hash_len,
+                 sig, &ck_sig_len );
+    if( rv != CKR_OK )
+        goto exit;
+
+    if( mechanism.mechanism == CKM_ECDSA )
+    {
+        /* The signature from the token contains r and s concatenated,
+         * each in the form of a big-endian byte sequence, with r and s
+         * having the same length as the base point.
+         *
+         * A standard ECDSA signature is encoded in ASN.1:
+         *   SEQUENCE {
+         *     r INTEGER,
+         *     s INTEGER
+         *   }
+         *
+         * Perform the conversion using existing utility functions,
+         * with temporary bignums.
+         */
+        uint16_t byte_len = ( ( ctx->bit_length + 7 ) / 8 );
+        size_t sig_size = MBEDTLS_ECDSA_MAX_SIG_LEN( ctx->bit_length );
+        mbedtls_mpi r, s;
+        mbedtls_mpi_init( &r );
+        mbedtls_mpi_init( &s );
+        rv = CKR_OK;
+        if( ck_sig_len != 2 * byte_len )
+        {
+            /* Bad data from the token */
+            rv = CKR_GENERAL_ERROR;
+            goto ecdsa_exit;
+        }
+        if( mbedtls_mpi_read_binary( &r, sig, byte_len ) != 0 ||
+            mbedtls_mpi_read_binary( &s, sig + byte_len, byte_len ) != 0 )
+        {
+            rv = CKR_HOST_MEMORY;
+            goto ecdsa_exit;
+        }
+        /* The signature buffer is guaranteed to have enough room for
+           the encoded signature by the pk_sign interface. */
+        if( ecdsa_signature_to_asn1( &r, &s, sig, sig_len, sig_size ) != 0 )
```

Comment:
I'd prefer to avoid the use of temporary MPI's here and instead work in place, directly calling a function which converts an ECDSA signature given as the concatenation of `r,s` (bigendian) into ASN.1. I think we should avoid MPI's unless we want to do arithmetic.

Reply:
Done as a new function - please re-review.